model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java

```diff
@@ -68,18 +68,48 @@
      * @throws IllegalStateException If {@code criteria} is not compatible, i.e. has not been originally created
      *   by the {@link #createCriteriaBuilder()} method of this object.
      */
-    Stream<V> read(ModelCriteriaBuilder<M> criteria);
+    default Stream<V> read(ModelCriteriaBuilder<M> criteria) {
+        return read(criteria, null);
+    }
+
+    /**
+     * Returns stream of objects satisfying given {@code criteria} from the storage.
+     * The criteria are specified in the given criteria builder based on model properties.
+     *
+     * @param criteria Criteria filtering out the object, originally obtained
+     *   from {@link #createCriteriaBuilder()} method of this object.
+     *   If {@code null}, it returns an empty stream.
+     * @param queryParameters parameters for the query like firstResult, maxResult, requested ordering, etc.
+     * @return Stream of objects. Never returns {@code null}.
+     * @throws IllegalStateException If {@code criteria} is not compatible, i.e. has not been originally created
+     *   by the {@link #createCriteriaBuilder()} method of this object.
+     */
+    Stream<V> read(ModelCriteriaBuilder<M> criteria, QueryParameters<M> queryParameters);
```

Comment:
Would something like this be possible?
```java
import static QueryParameters.*;

read(withCriteria(mcb));
read(withCriteria(mcb).paginate(offset, limit, CLIENT_ID));
read(withCriteria(mcb).orderBy(ID));
```

Reply:
It would be possible but this would require probably some more discussion. 

This way we would have some object that is wrapping `QueryParameters` together with `ModelCriteriaBuilder`. Isn't this basically the same as having QueryParameters stored directly in the `ModelCriteriaBuilder`? 

It would end up like this:
```java
store.createCriteriaBuilder()
    .compare(aField, EQ, expectedValue)
    .pagination(first, max, orderByField)
```

Reply:
The pagination must not be available for individual criteria. What would be the meaning of
```java
  store.createCriteriaBuilder()
    .or(store.createCriteriaBuilder() .compare(aField, EQ, expectedValue) .pagination(first, max, orderByField),
         store.createCriteriaBuilder() .compare(aField2, EQ, expectedValue2) .pagination(first, max, orderByField2))
    .pagination(0, 1)
```
?

Pagination and order is _query_ parameter, not criteria parameter. Thus it cannot be incorporated in the `ModelCriteriaBuilder`.